Feature/ctv 2728 initial raf true x setup  (WIP)

addisonm opened the pull request:
<del>Initial work in progress.  Things are a bit of a mess.  But the meat of this is contained in PlaybackTask for now.  I have some intentions to clean it up as more of it is fleshed out.  I originally had a vision of making a WIP branch and adding PRs to that, so its pointing to a second branch I created for now.</del>

This should be relatively fleshed out now.  Currently leaning on leaving everything colocated in the single task due to Roku design. Could move some things like the event handler and high level behaviours out if we wanted.

If there are any comments, I am all ears.

As a summary of whats integrated/working
- RAF integrations.  TrueX and non Truex Ad use cases
- TAR integrations.  Choice card, Skip card, Ad Interaction
- Some RAF Ad parsing

Main remaining items:
- Create AdParameter local stubs 
- Verify full ad experience with proper vast samples

addisonm committed:
CTV-2727
- Initial commit for work towards RAF/Truex reference app integration
- Introduce RafContentFlow to mirror ContentFlow as entry point to RAF implementation
- Create a task to setup RAF (RAF must be in a task or main thread)
-- RAF parses initial ad payload and detects ads per second
-- Introduce (temporarily hacky) handling to detect truex vs regular ads
-- Render basic TrueX ad experience.  Start, Interaction, and Exits (no complete yet)
- Include some sample ad payloads

addisonm committed:
CTV-2727
- Introduce RafContentFlow to keep existing flow separate for now
- Split off choice card cancel and interaction cancel events
- Add event bubbling and exit flow to restart video experience
- Minor change to truex ad detection

addisonm committed:
CTV-2727
- Update ad sample to simulate truex with followup ads if user skips
- Handle user opting out of choice card better
- Add some preliminary back handling to improve test experience

addisonm committed:
CTV-2727
- A lot of small improvements
- Introduced successful adfreepod handling (skip regular ads)
- Improve cleanup related flows
- Code clean up

addisonm committed:
CTV-2727
- Add better preroll handling for the first ad (but not subsequent prerolls)
- Clean up a lot of unnecessarily complexity
- Change truex local payload and to have a choice card timer
- Add a placeholder skip card to local vast payload

kylelamx commented on components/PlaybackTask.brs:
```diff
@@ -0,0 +1,244 @@
+  videoContent = CreateObject("roSGNode", "ContentNode")
+
+  ' videoContent.url = "http://development.scratch.truex.com.s3.amazonaws.com/roku/simon/roku-reference-app-stream-med.mp4"
+  videoContent.url = "http://commondatastorage.googleapis.com/gtv-videos-bucket/sample/ForBiggerFun.mp4"
```
you can use this:
http://ctv.truex.com.s3.amazonaws.com/assets/reference-app-stream-no-ads-720p.mp4

addisonm replied:
Done.  I'll leave the other stream for now because it loads much faster.  But will look to remove it for the final delivery.

kylelamx commented on components/PlaybackTask.brs:
```diff
@@ -0,0 +1,244 @@
+    if msgType = "roSGNodeEvent"
+      field = msg.getField()
+      ? "roSGNodeEvent msg.getField()" field
+      if field = "position" then 
+        onPositionChanged(msg)
```
I am not too familiar with the `while(true) msg = Wait(0, m.port)` but I think what we want here is just to observe the position change? If so, maybe we can use the `observeFieldScoped` like here:
https://github.com/socialvibe/truex-roku-reference-app/blob/master/components/ContentFlow.brs#L335

we can also change the interval with `notificationInterval` as mentioned here 
https://developer.roku.com/docs/references/scenegraph/media-playback-nodes/video.md#trickplay-fields

addisonm replied:
the `wait(0, m.port)` essentially listens for any events that subscribes to the port.  So it is a central way of handling any event based behaviour.  I find it a bit more readable if you have a lot of events and more flexible when building something complex.  As it turns out, I might only care about two events for this because of the scope of the effort, so I wouldn't be against changing the event handlers to be more explicit.

addisonm replied:
Interestingly, using explicit function names for the event callbacks, the callbacks are run on the render thread.  Being on the render thread changes the scope that I want.  I need to remain in the scope of the task, so I'll need to leave it as is for now.

kylelamx commented on components/PlaybackTask.brs:
```diff
@@ -0,0 +1,244 @@
+      if field = "position" then 
+        onPositionChanged(msg)
+      else if field = "event" then
+        onTrueXEvent(msg)
```
like video, same for the truex event
https://github.com/socialvibe/truex-roku-reference-app/blob/master/components/ContentFlow.brs#L136

kylelamx commented on components/PlaybackTask.brs:
```diff
@@ -0,0 +1,244 @@
+  position = event.getData()
+  ' ? "onPositionChanged: " position
+
+  ads = m.raf.getAds(event)
```
I think getAds takes a msg as string, no?

addisonm replied:
Good question.  Their API docs say it is a string, but all examples of it showcase passing the position event.  The doc does describe this though, passing the position event/msg.  Given that it functions as intended and matches enough information, this should be fine.

kylelamx commented on components/PlaybackTask.brs:
```diff
@@ -0,0 +1,244 @@
+      ads.viewed = true
+      m.skipAds = false
+
+      ' Show choice card if needed?
```
TAR handles that. For publishers, it looks like a normal Truex session.

kylelamx commented on components/PlaybackTask.brs:
```diff
@@ -0,0 +1,244 @@
+        placement_hash: "74fca63c733f098340b0a70489035d683024440d" 'Placeholder
+      }
+
+      ads.ads.delete(0) ' Removes it from future ad handling for raf
```
Why are we removing that? If the ad pod is replayable, then we do want TrueX ads to appears again

addisonm replied:
Currently need to do this, otherwise for opt out type flows, RAF will attempt to play the TrueX ad with its own ad renderer, followed by the regular ads.  The desired flow is to ignore the TrueX ad, and have RAF play the other ads.

Crackle was doing this, and the reasons for it make sense.  But I do see your concern.  Replayability becomes a concern.  One way to handle this is to parse all the ads ahead of time, then introduce some kind of a mapping, so the information isn't lost.  I am of the opinion rewinding streams is relatively uncommon, in terms of replaying ads.  So this could be more of an edge case concern too.

kylelamx commented on components/PlaybackTask.brs:
```diff
@@ -0,0 +1,244 @@
+  types = {
+    "ADSTARTED": "adStarted",
+    "ADCOMPLETED": "adCompleted"
+    "ADFIRSTQUARTILE": "adFirstQuartile",
+    "ADSECONDQUARTILE": "adSecondQuartile",
+    "ADTHIRDQUARTILE": "adThirdQuartile",
+    "ADPOSITION": "adPosition",
```
don't think Truex actually have these events

kylelamx commented on components/PlaybackTask.brs:
```diff
@@ -0,0 +1,244 @@
+  if ads <> invalid AND ads.ads.count() > 0
+    firstAd = ads.ads[0] 'Assume truex can only be first ad
+
+    if m.skipAds then
+      ads.viewed = true
+      m.skipAds = false
+
+      ' Show choice card if needed?
+      return
+    end if
```
looks likes this skipAds logic is in the wrong place. (resume ads, perhaps?)

addisonm replied:
This is intended, I've changed the logic for this a bit.  PR is a bit stale.

addisonm committed:
CTV-2727
- Polish
- Support external adUrl to the task

addisonm committed:
CTV-2727
- Fix a scoping issue

addisonm committed:
CTV-2727
- Rework preroll handling so RAF non truex prerolls don't break playback
- Rework handleAds to tell us return if handled ad or not
- Clean up some of the truex event cases to behave as intended
- Fix some logic surrounding scoping and external events, particularly on exit.

addisonm committed:
CTV-2727
- Clean up
- Improve preroll and resume playback logic
- Store last position so video resumes properly when restarting player

addisonm committed:
CTV-2727
- Switch default to local TrueX ad experience

addisonm committed:
CTV-2728 (Note typo in ticket number previously)
- Remove hacky workaround to truex ad detection (only AdParameter tag supported)
- Clean-up resource files to make it easier to follow

addisonm commented on components/MainScene.brs:
```diff
@@ -46,7 +46,8 @@ sub onFlowEvent(event as object)
     ? "TRUE[X] >>> MainScene::onFlowEvent(trigger=";data.trigger;")"
 
     if data.trigger = "playButtonSelected" then
-        showFlow("ContentFlow")
```
Thoughts on removing the other content flow entirely for this reference app, files and all?

kylelamx replied:
should just remove ContentFlow if it wasn't being used, to simplify the app.

episcopus replied:
Yes, overall, please remove all (now) superfluous cruft, including test stream files / json's that pertain to the previous life of this repurposed demo app. Simplify the flow and remove indirections as much as possible so it's as braindead as possible to get the RAF flow. Our existing apps are not the best examples since we have a few layers I feel we could do without, for instance the thing that fetches the streaminfo for the app. I would love to get rid of that and just hard code the image and video, since we don't use the `vmap` in `reference-app-streams.json` anyway.

addisonm commented on res/adpod/truex-pod.xml:
```diff
@@ -24,7 +24,7 @@
                         <MediaFiles>
                             <MediaFile id="coffee-video-ad" delivery="progressive" width="3840" height="2160" type="video/mp4" bitrate="48610" scalable="true" maintainAspectRatio="true">
-                                <![CDATA[https://qa-media.truex.com/reference_apps/sample_ads/video/coffee-4k.mp4]]>
+                                <![CDATA[http://commondatastorage.googleapis.com/gtv-videos-bucket/sample/ForBiggerEscapes.mp4]]>
```
I need to change these to be truex feeds.  Noting that I had to remove the above because my device did not seem to like the 4k files.

rblaa replied:
are 4k ads a real thing? .mp4 4k videos will choke most networks and devices
they should not exist in ads in practice right?

addisonm replied:
I would say so.  But I suppose there's nothing stopping it from happening either.  Sounds like a client / ad tech problem to deal with though.

addisonm committed:
CTV-2728
- Remove commented code entirely
- Clean up

addisonm committed:
CTV-2728
- Adjust ad parameters to have proper preroll/midroll tags
- Clean up

addisonm committed:
CTV-2728
- Comment some debug logs

addisonm committed:
CTV-2728
- Slight rework to instead get the ad, and let the natural playback flow handle ad playback to centralize "play" case
- Added more flexibility and options for sample ad payloads

kylelamx commented on res/ad/skipcard-vastconfig.json:
```diff
@@ -0,0 +1,54 @@
+{
+  "tag_type": "skip",
+  "user": {
+    "id": "2D1CCD5A-D52A-427C-8CEF-9ADC359FA083"
+  },
```
let's remove all hardcode truex responses before we merge

episcopus replied:
+1

addisonm replied:
Hmm, I was planning to use this as a mechanism to allow simulating the skip card flow, given that there have been some difficulties getting skip cards to return from the ad back end.  But if we are going to forgo that entirely for now, then I am fine with this.

Curious, what is the reason for removing it?  The vendor would be able to see this by tracing the http calls on their own.

episcopus replied:
The skip card is just one of the possible formats returned by our ad server, and the behavior is completely encapsulated within the TruexAdRenderer (TAR). Hence demonstrating or exercising it as a flow is not actionable or valuable to a publisher integrating trueX which is what this project is focused on helping with.

kylelamx commented on res/adpod/truex-pod-preroll.xml:
```diff
@@ -0,0 +1,57 @@
+            <AdSystem>trueX</AdSystem>
+            <AdTitle>true[X] Interactive Ad</AdTitle>
+            <Description>Interactive ad from true[X]</Description>
+            <VASTAdTagURI><![CDATA[http://get.truex.com/f10e25d796a25c294dc308f2b5f916cdd3d670e9/vast?dimension_5=evergreen&stream_position=preroll&network_user_id=ROKU_ADS_TRACKING_ID]]></VASTAdTagURI>
```
should be:
```http://get.truex.com/f10e25d796a25c294dc308f2b5f916cdd3d670e9/vast?dimension_2=0&dimension_5=evergreen&stream_position=preroll&network_user_id=ROKU_ADS_TRACKING_ID```
I think you accidentally removed `dimension_2=0`

addisonm replied:
Done.

kylelamx commented on components/PlaybackTask.brs:
```diff
@@ -0,0 +1,253 @@
+    m.currentAdPod = ads
+    firstAd = ads.ads[0] 'Assume truex can only be first ad
+
+    if firstAd.adParameters <> invalid
```
we might want to do something like 
```
if isTrueXAd(firstAd) then
```
and `isTrueXAd` should check more than just `.adParameters <> invalid`, for example, if it has vast_config_url and if the URL match `.*get.truex.com.*`

episcopus replied:
Alternatively not sure if it comes through RAF but the following gets returned by the tag: `<AdSystem>trueX</AdSystem>`. But yeah if there's a way to make it a bit more selective w/o overburdening the code, I think this would be helpful.

addisonm replied:
True, the ads might not be from TrueX but use the same tags.  Let me look at whats returned by RAF here more closely.

addisonm replied:
`AdSystem` is not parsed by RAF unfortunately.

I opted for `adServer` property, to avoid having to dive into parsing the adparameters and figure that TAR will reject bad urls.  Assuming that all TrueX ad's will be sourced by `get.truex.com`.  For simplicity I opted not to use a regex.

rblaa replied:
you should support qa-get.truex.com as well

addisonm replied:
Good point.  The code happens to check it right now, but I'll make it more explicit.

kylelamx commented on components/PlaybackTask.brs:
```diff
@@ -0,0 +1,253 @@
+      m.truexAd = firstAd
+      m.truexAd.adParameters = parseJSON(m.truexAd.adParameters)
+      m.truexAd.renderSequence = ads.renderSequence
+      ads.ads.delete(0) ' Removes it from certain use cases due to raf
```
Would be awesome if the comment explains this better.

kylelamx commented on components/PlaybackTask.brs:
```diff
@@ -0,0 +1,253 @@
+    end if
+
+    ' Check if we need to play other (non-truex) ads
+    play = handleAds(m.currentAdPod)
```
how about... `resumePlayback` instead of `play`?

addisonm replied:
I was a bit conflicted with the name because this is also responsible for starting playback the first time.  Resume makes less sense for that initial case.  It didn't seem necessary to create a separate start playback flow when it is handled by one general flow.

Edit: Renamed the function to `playContentStream` if that helps.

kylelamx commented on components/PlaybackTask.brs:
```diff
@@ -0,0 +1,253 @@
+  adUrl = m.top.adUrl
+  if adUrl = invalid OR adUrl = ""
+    ' adUrl = "pkg:/res/adpods/vast-regular.xml"  ' Regular Ads
+    adUrl = "pkg:/res/adpods/vmap-truex.xml" ' Preroll + Midroll Truex experience
```
I think the Preroll + Midroll Truex experience should be the only one we keep after we are done with testing.

episcopus replied:
+1 remove the commented stuff if we don't need it

kylelamx commented on components/PlaybackTask.brs:
```diff
@@ -0,0 +1,253 @@
+  startPlaybackEvents = [types.ADCOMPLETED, types.NOADSAVAILABLE, types.ADERROR]
+
+  if eventType = types.ADFREEPOD
+    m.skipAds = true
```
you might need to reset m.skipAds to false on adStarted? else the midroll will be skipped?

addisonm replied:
Good point.  I was mentally using the Sponsored Stream case as my primary use case.  But this technically would break for non-sponsored stream cases.  Will take a closer look at this, have some simple ideas.

Edit: I now set the flag to false after setting the pod to viewed in `playContentStream`.  This should take care of the issue you raised.

kylelamx commented on components/PlaybackTask.brs:
```diff
@@ -0,0 +1,253 @@
+  videoContent.url = "http://ctv.truex.com.s3.amazonaws.com/assets/reference-app-stream-no-ads-720p.mp4"
+  videoContent.length = 22 * 60
+
+  videoContent.title = "video title"
```
this technically has a title of "The true[X] Employee Experience" :p
https://github.com/socialvibe/truex-roku-reference-app/blob/master/res/reference-app-streams.json#L3

addisonm replied:
Done.

kylelamx commented on components/PlaybackTask.brs:
```diff
@@ -0,0 +1,253 @@
+      else if field = "exitPlayback"
+        exitPlayback()
+      end if
+      ' TODO: Video complete event
```
I think you did that above already? Or this is something else?

addisonm replied:
Perhaps less of a TODO.  But I did not handle the "end of the video" case.  It felt out of scope related to ads and all it would do is go back to the detail page.  User should be able to just press back (I need to double check in case something weird happens on video end)

addisonm replied:
Scrubbed the comment to avoid confusion.  If this is desired behaviour for the ref app, let me know.  I figure it is listening for an event, or checking if the position is at the end, then exit the player.

episcopus commented on components/PlaybackTask.brs:
```diff
@@ -0,0 +1,253 @@
+Library "Roku_Ads.brs" ' Must be initialized in the task (or main) thread
+
+sub init()
+  ? "TRUE[X] >>> PlaybackTask::init()"
```
Can you update to 4 spaces for indentation everywhere to be consistent

episcopus commented on components/PlaybackTask.brs:
```diff
@@ -0,0 +1,253 @@
+
+sub setupScopedVariables()
+  m.port = createObject("roMessagePort")
+  m.adFacade = m.top.adFacade
```
Can you explain this whole `adFacade` deal? I was reviewing the flow and noticed this `<Rectangle>` being passed from `RafContentFlow.xml`. Is this a RAF requirement?

addisonm replied:
This is more for rendering the TrueX ad.  Similar to the player, that is passed up from the render thread to the task thread to allow indirect access to the UX.  This was inspired from the Crackle sample and is similar to the existing ref app appending it to the render thread directly.
```
` instantiate TruexAdRenderer and register for event updates
m.adRenderer = m.top.createChild("TruexLibrary:TruexAdRenderer") 
```

episcopus replied:
OK I guess it's fine if you believe it's a best practice and a good way to do it. I just thought it was odd to use a `<Rectangle>` instead of a `<Group>`, is this to mask the underlying UI / player as we spin up the interactive experience?

addisonm replied:
Group works fine (will changed to this).  I didn't concern myself too much with what the container was.

This is more to give something to attach the AdRenderer that comes from the render thread.  The player is handled by visibility state.

episcopus commented on components/PlaybackTask.brs:
```diff
@@ -0,0 +1,253 @@
+  initPlayback()
+end sub
+
+sub setupScopedVariables()
```
Please add comments documenting each function.

episcopus commented on components/PlaybackTask.brs:
```diff
@@ -0,0 +1,253 @@
+  videoPlayer = m.top.video
+  videoContent = createObject("roSGNode", "ContentNode")
+
+  videoContent.url = "http://ctv.truex.com.s3.amazonaws.com/assets/reference-app-stream-no-ads-720p.mp4"
```
Url can be shorter: `http://ctv.truex.com/assets/reference-app-stream-no-ads-720p.mp4`

episcopus replied:
If we hard code everything here (which is fine) and as per my above comment, can we remove all the stream info loading cruft?

episcopus commented on components/PlaybackTask.brs:
```diff
@@ -0,0 +1,253 @@
+
+  ? "TRUE[X] >>> PlaybackTask::onTrueXEvent(): " eventType
+
+  types = {
```
Seems a lot of effort to set up and call into a utility function to simply test the `eventType` against 3 events. I think it's better to just OR this in a single line to remove extraneous code.

episcopus commented on components/PlaybackTask.brs:
```diff
@@ -0,0 +1,253 @@
+
+  if eventType = types.ADFREEPOD
+    m.skipAds = true
+  else if arrayUtils_includes(startPlaybackEvents, eventType)
```
Would a better name for this be something along the lines of `resumeContentStream` and `exitContentStream`? I think this helps clarify it's the underlying content stream that's going to be manipulated, not the trueX ads.

episcopus commented on components/PlaybackTask.brs:
```diff
@@ -0,0 +1,253 @@
+      playTrueXAd()
+      resumePlayback = false
+    else ' Non-TrueX ads
+      hidePlayback()
```
Similar comment here, naming to `hideContentStream()` could be more helpful.

episcopus commented on components/RafContentFlow.brs:
```diff
@@ -0,0 +1,41 @@
+    ? "TRUE[X] >>> RafContentFlow::initRafTask()"
+
+    if m.rafTask = invalid then
+        rafTask = CreateObject("roSGNode", "PlaybackTask")
```
I'm fuzzy here, but does this really all need to be wrapped in a task?

addisonm replied:
Unfortunately yes.  That or on the main thread.  Reason being a RAF requirement.  It will just throw some error if you don't follow this.

episcopus commented on res/ad/choice-card.xml:
```diff
@@ -0,0 +1,27 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
What is this file about?

addisonm replied:
I've removed this (it was also missing a commit in progress to make use of it).  They were to create stubbed ad experiences but they are just fluff.

episcopus commented on res/ad/skip-card.xml:
```diff
@@ -0,0 +1,27 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Is this a test file?

episcopus commented on res/adpod/truex-pod-skip.xml:
```diff
@@ -0,0 +1,57 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Is this needed?

addisonm replied:
Removed

episcopus commented on res/adpods/vmap-truex.xml:
```diff
@@ -0,0 +1,15 @@
+<vmap:VMAP xmlns:vmap="http://www.iab.net/videosuite/vmap" version="1.0">
```
Just curious, but can RAF support being instantiated with the type of VAST in `truex-pod-preroll.xml` or `truex-pod-midroll.xml` instead?

The way the client described to us is they make a ad call right before every pod (I assume they hard code their pod break index or comes in a separate CMS?) and get back VAST. They explicitly stated not working with VMAP. 

So I was curious if there's a way to go directly to VAST or RAF simply had to be instantiated with VMAP.

I'm not asking for any change at this stage, as I think the example is readable even with VMAP, and can readily be adapted, just trying to understand in case there's pushback on the use of VMAP in the example...

addisonm replied:
The vast file's can be passed directly (eg truex-pod-preroll.xml) and it will work.  I didn't have a clear understanding of what OTTera's ad layer was like.  But given that you can pass the pods immediately, I figured it will be fine to showcase the flow with vmap and then adjust from there.  (Also had some difficulty with the ad construction layer)

Your description does sound like a bit of a gap though.  If the client is going to make ad calls immediately before each pod, it makes part of the RAF logic a bit odd.  They would need to replace raf.getAds() with their own behaviour, along with some kind of ad position mechanism.  I suppose one can assume they should have some of this in place already, they just need to adjust it to support TrueX.....

episcopus replied:
You are making good points. All things considered let's stick with the VMAP mechanism for this version and take it from here with the client.

episcopus commented on res/reference-app-streams.json:
```diff
@@ -1,27 +1,27 @@
-            }
-        ]
-    }
+  {
```
No longer needed right?

episcopus commented on source/Utils.brs:
```diff
@@ -0,0 +1,17 @@
+' Some general helper functions 
+'-----------------------------------------------------
+
+function arrayUtils_includes(array, value)
```
We can get rid of this file

addisonm replied:
Done

episcopus reviewed (changes requested):
Overall looks pretty good, excited to see a RAF demo and one that looks pretty simple at that.

General feedback is this needs to be cleaned up quite a bit, documented, also, with clear and simple instructions in the `README.md` file, pointing to the salient parts of the flow. There's a lot of indirections no longer essential here which we can take out.

There's a few things I was confused about which possibly betray a lack of familiarity with RAF from my end.

addisonm committed:
CTV-2728
- First round of cleanup PR feedback

addisonm committed:
CTV-2727
- More cleanup

addisonm committed:
CTV-2728
- More clean up

addisonm committed:
CTV-2728
- More cleanup

addisonm committed:
CTV-2727
- Fix an issue where midroll ads could get skipped undesirably

addisonm committed:
CTV-2728
- Adjust name to playContentStream to make it start/resume incognizant

addisonm committed:
CTV-2728
- Improve TrueX ad detection to be more restrictive

addisonm committed:
CTV-2728
- Clean up spacing standards

addisonm committed:
CTV-2728
- Remove m.truexAd variable and just pass it as a parameter where needed instead
- Added documentation for functions

addisonm committed:
CTV-2728
- Added more comments

addisonm committed:
CTV-2728
- Remove some unnecessary files for now for simplicity.

addisonm committed:
CTV-2728
- Clean up
- Remove some AWS references

addisonm committed:
CTV-2728
- Update Readme

addisonm committed:
CTV-2728
- Formatting

addisonm committed:
CTV-2728
- Explicitly check for prod and qa domains for the adserver for better detection (currently overlaps, but for clarity)

episcopus commented on README.md:
```diff
@@ -1,15 +1,15 @@
 # Implementation Details
 
-In this project we simulate the integration with a live Ad Server (CSAI use case) or SSAI provider through a mock ad playlist configuration. This is meant to capture the stream's ad pods, including their duration and the reference to the true[X] payloads in each pod. This configuration is maintained in `res/reference-app-streams.json` as part of the `vmap` key. In this sample channel, two ad brea ...
+For the purpose of this project, some of the data is hard coded as the primary goal here is to showcase the RAF - TrueX Ad integration for client side handled ads. The core of this implementation starts in `PlaybackTask`. As a brief summary, the app will launch and prepare the TrueX Ad Renderer (TAR). Once ready, the app will launch to a simple page (`DetailsFlow`), and waits for the user to star ...
```
Should be more like "As a brief summary, the app will **load** and prepare the TrueX Ad Renderer (TAR) **component library**. Once ready..."

The way it was worded makes it sound like a TAR instance is created on channel launch or super early in the app lifecycle while actually it's right in line with processing the results of ad requests. I think you meant to speak to the component library load actually so revising this makes sense.

episcopus commented on README.md:
```diff
@@ -1,15 +1,15 @@
+For the purpose of this project, some of the data is hard coded as the primary goal here is to showcase the RAF - TrueX Ad integration for client side handled ads. The core of this implementation starts in `PlaybackTask`. As a brief summary, the app will launch and prepare the TrueX Ad Renderer (TAR). Once ready, the app will launch to a simple page (`DetailsFlow`), and waits for the user to star ...
 
-This `vmap` ad playlist is marshaled through to the `ContentFlow` SceneGraph Component which handles the stream playback. In `preprocessVmapData` we parse out this ad playlist and build simple data structures which are then referenced as part of the video position change handler (`onVideoPositionChange`) to detect when we encounter a true[X] ad pod. We then initialize and launch the true[X] ad wi ...
+In the task, we initialize the player with video metadata and RAF with ads. A very basic use case, wrapped in `vmap-truex.xml` is the default ad payload, with a preroll and midroll, each containing a TrueX ad. The payload is meant to be a simulation of a valid ad payload from the ad server. While the preroll and midroll are also local files, the TrueX ads are real server side ads, and can be seen ...
```
I would reword a bit to make it clearer that the `vmap-truex.xml` is a proxy / mock for a live ad server response and a real world app would simply substitute in its ad server URL / tag. RAF does the rest (abstracting away the details / specifics of the ad server being used) and the rest of the RAF flow should work as demonstrated.

episcopus commented on README.md:
```diff
@@ -1,15 +1,15 @@
+In the task, we initialize the player with video metadata and RAF with ads. A very basic use case, wrapped in `vmap-truex.xml` is the default ad payload, with a preroll and midroll, each containing a TrueX ad. The payload is meant to be a simulation of a valid ad payload from the ad server. While the preroll and midroll are also local files, the TrueX ads are real server side ads, and can be seen ...
 
-The `onTruexEvent` subroutine handles true[X] events and repositions the playhead depending on whether the viewer completed an ad or not (earning an ad pod skip in the former case). 
+The core of ad handling is managed by `handleAds()`. This takes care of determining if there is a valid ad to play from `raf.getAds()`. For this sample, it supports the `<AdParameter>` tag to designate a TrueX Ad Pod. `playTrueXAd` will take care of playing a given TrueX ad pod rendered on the `adFacade` that was passed from the flow. The TAR event listener `onTruexEvent` will take care of the us ...
```
As with the rest of the README, this reads very nice! 👍 

I would reword the third sentence like so: "We check for the presence of `<AdParameters>` and `get.truex.com` in the returned URL to determine whether a true[X] ad is present."

episcopus commented on components/DetailsFlow.brs:
```diff
@@ -34,8 +34,12 @@ sub init()
 
-    if m.global.streamInfo = invalid then return
-    streamInfo = ParseJson(m.global.streamInfo)[0]
+    streamInfo = {
```
Yes, thank you!

episcopus commented on components/MainScene.brs:
```diff
@@ -25,9 +25,6 @@ sub init()
 
     ' create/set global fields with Channel dimensions (m.global.channelWidth/channelHeight)
     setChannelWidthHeightFromRootScene()
```
Removing the `LoadingFlow` who then takes care of properly loading the component library? I think showing a pattern of how to load the component library (and ensure it's properly completed loading before doing anything with true[X]) is a useful pattern, but admittedly non essential. If you feel we can do away with it and simplify the example and still make it work, I'm ok with this.

addisonm replied:
Oddly enough, this pattern is taken care of in the main scene.  Loading flow had nothing to do with the component library.  It would be nice to include the spinner, but thats not the focus of the app.

episcopus commented on components/PlaybackTask.brs:
```diff
@@ -1,253 +1,348 @@
+    adUrl = m.top.adUrl
+    if adUrl = invalid OR adUrl = ""
+        adUrl = "pkg:/res/adpods/vmap-truex.xml" ' Preroll + Midroll Truex experience
+        ' adUrl = "pkg:/res/adpod/truex-pod-preroll.xml"  ' Can check individual vast pods
```
Cool, but if you are going to leave this line in here, I assume we tested this alternate xml file and it works? 😅

addisonm replied:
Yep :).  The caveat is Roku always treats it as a preroll, which makes sense since you didn't provide contextual time information.

episcopus commented on components/PlaybackTask.brs:
```diff
@@ -1,253 +1,348 @@
+end function
+
+'-------------------------------------------
+' Helper to see if an ad is TrueX
```
Add more color here that explains that our standard ad tag includes `<AdParameters>` and points to a get.truex.com DNS.

episcopus reviewed (approved):
lgtm!

addisonm committed:
CTV-2728
- Clean up readme / comments per PR comments

kylelamx reviewed (approved):
LGTM
